systemverilog-plugin/UhdmAst.cc

```diff
@@ -1891,11 +1898,39 @@ void UhdmAst::process_module()
                     }
                 }
                 if (shared.top_nodes.count(type)) {
-                    if (!node->children[0]->str.empty())
+                    if (!node->children[0]->str.empty()) {
                         module_parameters += node->str + "=" + node->children[0]->str;
-                    else
+                    } else if (node->children[0]->bits.size() > 32) {
+                        std::string value = std::to_string(node->children[0]->bits.size()) + "'b";
+                        for (const auto &bit : node->children[0]->bits) {
+                            switch (bit) {
+                            case RTLIL::State::S0:
+                                value += "0";
+                                break;
+                            case RTLIL::State::S1:
+                                value += "1";
+                                break;
+                            case RTLIL::State::Sx:
+                                value += "x";
+                                break;
+                            case RTLIL::State::Sz:
+                                value += "z";
+                                break;
+                            case RTLIL::State::Sa:
+                                value += "a";
+                                break;
+                            case RTLIL::State::Sm:
+                                value += "m";
+                                break;
+                            default:
+                                report_error("Unhandled RTLIL State case!\n");
+                            }
+                        }
+                        module_parameters += node->str + "=" + value;
+                    } else {
                         module_parameters +=
                           node->str + "=" + std::to_string(node->children[0]->bits.size()) + "'d" + std::to_string(node->children[0]->integer);
+                    }
```

Comment:
Shouldn't it follow the way Yosys does it in [`serialize_param_value`][serialize_param_value]?

I think you can even replace this whole snippet (and a bit above and below) with [`AstNode::derived_module_name`][derived_module_name] + [`AstNode::asParaConst`][asParaConst].

[derived_module_name]: https://github.com/YosysHQ/yosys/blob/7c5ae560a838db3f35dec0760df6cd8ebb590b80/frontends/ast/ast.h#L386
[asParaConst]: https://github.com/YosysHQ/yosys/blob/7c5ae560a838db3f35dec0760df6cd8ebb590b80/frontends/ast/ast.h#L319
[serialize_param_value]: https://github.com/YosysHQ/yosys/blob/7c5ae560a838db3f35dec0760df6cd8ebb590b80/frontends/ast/ast.cc#L1664-L1685

Reply:
I didn't know about this functions, I've extracted this change to separate PR: https://github.com/chipsalliance/yosys-f4pga-plugins/pull/466